File: String/067. Add Binary.cpp
/*
67. Add Binary

Given two binary strings, return their sum (also a binary string).

The input strings are both non-empty and contains only characters 1 or 0.

Example 1:

Input: a = "11", b = "1"
Output: "100"
Example 2:

Input: a = "1010", b = "1011"
Output: "10101"

*/

class Solution {
public:
    string addBinary(string a, string b) {
        int i = a.size() - 1, j = b.size() -1, carry = 0;
        vector<int>vec;
        while (i >= 0 || j >= 0 || carry >0){
            if(i>=0)  carry += ( a[i] - '0');
            if(j>= 0) carry += (b[j] - '0');
            vec.push_back(carry%2);//生成的是反的顺序的
            carry >>=1 ;
            i--; j--;
        }
        string res;
        transform(vec.rbegin(), vec.rend(), back_inserter(res), [](const int & i){return i+'0';});
        return res;
    }
};

class Solution {
public:
    string addBinary(string a, string b) {
        int i = a.size() - 1, j = b.size() -1, carry = 0;
        string res;
        vector<int>vec;
        while (i >= 0 || j >= 0 || carry >0){
            if(i>=0)  carry += ( a[i] - '0');
            if(j>= 0) carry += (b[j] - '0');
            res+=to_string(carry%2);//生成的是反的顺序的
            carry >>=1 ;
            i--; j--;
        }
        reverse(res.begin(),res.end());
        return res;
    }
};

File: String/067. Add Binary.py
"""
67. Add Binary

"""


class Solution:
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        i = len(a)-1; j = len(b)-1
        carry = 0
        res = ""
        while i >= 0 or j >= 0 or carry > 0:
            carry += ord(a[i]) - ord('0') if i >= 0 else 0
            carry += ord(b[j]) - ord('0') if j >= 0 else 0
            res+=str(carry%2)
            carry >>= 1
            i-=1
            j-=1
        return res[::-1]
            
class Solution:
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        ## convert the number a and b into interger with base 2
        ## convert the sum of a and b into the binary string
        return bin(int(a,2) + int(b,2))[2:]

class Solution:
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        return '{0:b}'.format(int(a, 2) + int(b, 2))

   class Solution:
        def addBinary(self, a, b):
            if len(a)==0: return b
            if len(b)==0: return a
            if a[-1] == '1' and b[-1] == '1':
                return self.addBinary(self.addBinary(a[0:-1],b[0:-1]),'1')+'0'
            if a[-1] == '0' and b[-1] == '0':
                return self.addBinary(a[0:-1],b[0:-1])+'0'
            else:
                return self.addBinary(a[0:-1],b[0:-1])+'1'